### PR TITLE
`MockBackend`: Implement later-stage witness generation

### PR DESCRIPTION
**Pull request by georgwiese**
Ticks the third item off #2152

This should make the mock prover feature-complete!

### CODE REVIEW

File: backend/src/mock/mod.rs [hunk 1]
```diff
@@ -70,37 +80,63 @@ impl<F: FieldElement> Backend<F> for MockBackend<F> {
...
+                    pil,
+                    &witgen_callback,
+                    &challenges,
+                )
             })
             .map(|machine| (machine.machine_name.clone(), machine))
             .collect::<BTreeMap<_, _>>();
 
         let mut is_ok = true;
         for (_, machine) in machines.iter() {
```
Comment (Schaeff) [inline backend/src/mock/mod.rs:127]:
```suggestion
        for machine in machines.values() {
```

File: backend/src/mock/mod.rs [hunk 2]
```diff
@@ -70,37 +80,63 @@ impl<F: FieldElement> Backend<F> for MockBackend<F> {
...
-            let result = PolynomialConstraintChecker::new(machine).check();
+            let result = PolynomialConstraintChecker::new(machine, &challenges).check();
             is_ok &= !result.has_errors();
         }
```
Comment (Schaeff) [inline backend/src/mock/mod.rs:130]:
```
let is_ok = machines.values().all(|machine| !PolynomialConstraintChecker::new(machine, &challenges).check().has_errors());
```

**Review by Schaeff — COMMENTED** [review]
LGTM except small form comments. Do we have enough negative tests?

**Comment by georgwiese** [comment]
> Do we have enough negative tests?

No, we have zero!

**Pull request merged**